library/psa_crypto.c

```diff
@@ -3131,7 +3156,733 @@ psa_status_t psa_asymmetric_decrypt( mbedtls_svc_key_id_t key,
+        * defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) &&
+        * defined( MBEDTLS_ECP_RESTARTABLE ) */
+}
+
+psa_status_t mbedtls_psa_sign_hash_abort(
+                    mbedtls_psa_sign_hash_interruptible_operation_t *operation )
+{
+
+#if (defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) || \
+     defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) ) && \
+     defined( MBEDTLS_ECP_RESTARTABLE )
+
+    if( operation->ctx )
+    {
+        mbedtls_ecdsa_free( operation->ctx );
+        mbedtls_free( operation->ctx );
+        operation->ctx = NULL;
+    }
+
+    mbedtls_ecdsa_restart_free( &operation->restart_ctx );
+
+    return PSA_SUCCESS;
+
+#else
+
+    ( void ) operation;
+
+    return( PSA_ERROR_NOT_SUPPORTED );
+
+#endif /* defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) ||
+    * defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) &&
+    * defined( MBEDTLS_ECP_RESTARTABLE ) */
+}
+
+psa_status_t mbedtls_psa_verify_hash_start(
+                mbedtls_psa_verify_hash_interruptible_operation_t *operation,
+                const psa_key_attributes_t *attributes,
+                const uint8_t *key_buffer, size_t key_buffer_size,
+                psa_algorithm_t alg,
+                const uint8_t *hash, size_t hash_length,
+                const uint8_t *signature, size_t signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    size_t coordinate_bytes = 0;
+    size_t required_hash_length = 0;
+
+    if( !PSA_KEY_TYPE_IS_ECC( attributes->core.type ) )
+        return PSA_ERROR_NOT_SUPPORTED;
+
+    if( !PSA_ALG_IS_ECDSA( alg ) )
+        return PSA_ERROR_NOT_SUPPORTED;
+
+#if (defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) || \
+     defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) ) && \
+     defined( MBEDTLS_ECP_RESTARTABLE )
+
+    /* Ensure default is set even if
+     * mbedtls_psa_interruptible_set_max_ops() has not been called. */
+    mbedtls_psa_interruptible_set_max_ops(
+                           mbedtls_psa_interruptible_get_max_ops( ) );
+
+    status = mbedtls_psa_ecp_load_representation( attributes->core.type,
+                                                  attributes->core.bits,
+                                                  key_buffer,
+                                                  key_buffer_size,
+                                                  &operation->ctx );
+
+    if( status != PSA_SUCCESS )
+        return status;
+
+    coordinate_bytes = PSA_BITS_TO_BYTES( operation->ctx->grp.pbits );
+
+    if( signature_length != 2 * coordinate_bytes )
+        return PSA_ERROR_INVALID_SIGNATURE;
+
+    mbedtls_mpi_init( &operation->r );
+    status = mbedtls_to_psa_error(
+                   mbedtls_mpi_read_binary( &operation->r,
+                                            signature,
+                                            coordinate_bytes ) );
+
+    if( status != PSA_SUCCESS )
+        return status;
+
+    mbedtls_mpi_init( &operation->s );
```

Comment:
The placement of the various `_init()` calls (`r`, `s`, then `restart_ctx`) does not look correct to me. I'd expect all `_init()` calls at the beginning, or perhaps better happening before this function is even called, but not interleaved with calls that can fail.

Having a return statement followed by an `init()` call is usually a bad smell, as it makes it likely that we reach a point where we try to clear the structure due to an error, and not all members have been initialized yet, which would result in memory errors.

Am I missing something?

Reply:
My thoughts here were merely to try and avoid having to do work that would not be required, thus placed the inits after the early error out checks. The early error checks do not branch to exit, merely return the error, so the the frees would not happen either.

If you think this is going to create maintenance burden I can change it, but this was why I did it this way.

Reply:
> The early error checks do not branch to exit, merely return the error, so the the frees would not happen either.

I think the frees (1) need to happen in order to avoid leaking memory in case of errors (for example, if we fail to read `s`, we need to free `r`'s memory), and (2) are already happening because if this function returns an error our caller will end up calling `mbedtls_psa_verify_hash_abort()`. So no leak, but now the next problem is, we don't want to end up calling `free()` on something uninitialized, which would even worse than leaking memory.

I we're relying on the `init()` calls here to ensure that, then we're in trouble: if we get an error reading `r`, we'll return and eventually end up in the `abort()` function freeing `s` before we've called `init()`. To avoid that, we want all the `init()` functions called before anything can fail, return and eventually end up freeing stuff. That's why I was saying that `init()` after `return` is a bad smell.

Now in that case in practice I think the code is actually safe, because before calling `start()`, the user must initialize the operation structure in one of the legal ways, which all end up setting all substructures (`r`, `s` and `restart_ctx`) to `{0}` which is practice is equivalent to calling `init()` for those structure, even though I don't think that's documented.

So, bottom line is: I think the code is actually safe as it is (as far as memory management is concerned) but _looks_ like it isn't, and so I think we should change it. I can see two ways:
1. Drop the calls to `init()` here and just make sure `MBEDTLS_VERIFY_SIGN_HASH_INTERRUPTIBLE_OPERATION_INIT` does the same job.
2. Or keep the `init()` call but put them all at the top.

I think the first way is cleaner, but checking the documentation of `mbedtls_mpi`, it appears the current initalizer of `{0}` is not correct: the documentation [explicitly says so](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/bignum.h#L214)! (I didn't know, actually I thought `{0}` was correct and was merely checking if that was documented or not...) So I think we should go with option 1 but fix `MBEDTLS_VERIFY_SIGN_HASH_INTERRUPTIBLE_OPERATION_INIT` to correctly initiliaze MPIs. Probably the best way to do that is by creating `MBEDTLS_MPI_INIT` in `bignum.h`, and similarly for `restart_ctx` in `ecdsa.h`.

Hope it makes sense. Wdyt?

Reply:
Have gone for option 2, as after discussion with Janos, I am not sure an initialiser for an MPI is correct (and things are due to change on that front very shortly.) I have moved all of the init calls as early as possible in both sign and verify. Note that initialising an MPI to zero is safe for freeing it, its just not safe for use, as the sign will be indeterminate. Likewise the restart_ctx is safe to be free'd after zeroing.

Reply:
An initializer for an mpi would be correct, but I'd rather not do it, both because this is really not the place to change the bignum API and because bignum is on its way out so we shouldn't be extending it much.